sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,588 @@
+ * {@code ClientId}, a {@code ServerURI}, a {@code Topic} pattern, and optionally {@code
+ * username} and {@code password} to connect to the MQTT broker. The following
+ * example illustrates various options for configuring the source:
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(
+ *   MqttIO.read()
+ *    .withConnectionConfiguration(MqttIO.ConnectionConfiguration.create(
+ *      "tcp://host:11883",
+ *      "my_topic"))
+ *
+ * }</pre>
+ *
+ * <h3>Writing to a MQTT broker</h3>
+ *
+ * <p>MqttIO sink supports writing {@code byte[]} to a getTopic on a MQTT broker.
+ *
+ * <p>To configure a MQTT sink, as for the read, you have to specify a MQTT connection
+ * configuration with {@code ServerURI}, {@code Topic}, ...
+ *
+ * <p>The MqttIO only fully supports QoS 1 (at least once). It's the only QoS level guaranteed
+ * due to potential retries on bundles.
+ *
+ * <p>For instance:
+ *
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...) // provide PCollection<byte[]>
+ *   .MqttIO.write()
+ *     .withConnectionConfiguration(MqttIO.ConnectionConfiguration.create(
+ *       "tcp://host:11883",
+ *       "my_topic"))
+ *
+ * }</pre>
+ */
+public class MqttIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MqttIO.class);
+
+  public static Read read() {
+    return new AutoValue_MqttIO_Read.Builder()
+        .setMaxReadTime(null).setMaxNumRecords(Long.MAX_VALUE).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_MqttIO_Write.Builder()
+        .setRetained(false)
+        .build();
+  }
+
+  private MqttIO() {
+  }
+
+  /**
+   * A POJO describing a MQTT connection.
+   */
+  @AutoValue
+  public abstract static class ConnectionConfiguration implements Serializable {
+
+    @Nullable abstract String getServerUri();
+    @Nullable abstract String getTopic();
+    @Nullable abstract String getClientId();
+    @Nullable abstract String getUsername();
+    @Nullable abstract String getPassword();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setServerUri(String serverUri);
+      abstract Builder setTopic(String topic);
+      abstract Builder setClientId(String clientId);
+      abstract Builder setUsername(String username);
+      abstract Builder setPassword(String password);
+      abstract ConnectionConfiguration build();
+    }
+
+    /**
+     * Describe a connection configuration to the MQTT broker. This method creates an unique random
+     * MQTT client ID.
+     *
+     * @param serverUri The MQTT broker URI.
+     * @param topic The MQTT getTopic pattern.
+     * @return A connection configuration to the MQTT broker.
+     */
+    public static ConnectionConfiguration create(String serverUri, String topic) {
+      checkArgument(serverUri != null,
+          "MqttIO.ConnectionConfiguration.create(getServerUri, getTopic) called with null "
```

Comment:
Please fix this. It should be `"MqttIO.ConnectionConfiguration.create(serverUri, topic) called with null serverUri` etc. throughout the PR.